lib/src/signal_client.dart

```diff
@@ -48,31 +38,77 @@ class SignalClient {
 
   bool get connected => _connected;
 
-  Future<void> join(String url, String token, JoinOptions? options) async {
-    final rtcUrl = '$url/rtc';
-    var params = _joinParams(token);
-    if (options != null && options.autoSubscribe != null) {
-      params += '&auto_subscribe=${options.autoSubscribe! ? '1' : '0'}';
-    }
+  Uri _buildUri(
+    String uriString, {
+    required String token,
+    ConnectOptions? options,
+    bool reconnect = false,
+    bool validate = false,
+  }) {
+    final uri = Uri.parse(uriString);
+
+    return uri.replace(
+      //
+      // It is possible to enforce WSS
+      //
+      scheme: validate ? 'http' : uri.scheme,
+      path: validate ? 'validate' : 'rtc',
+      queryParameters: <String, String>{
+        'access_token': token,
+        'protocol': protocolVersion.toString(),
+        if (options != null) 'auto_subscribe': options.autoSubscribe ? '1' : '0',
+        if (reconnect) 'reconnect': '1',
+      },
+    );
+  }
+
+  Future<void> join(
+    String url,
+    String token, {
+    ConnectOptions? options,
+  }) async {
+    //
+    // Create default options if null
+    //
+    options ??= const ConnectOptions();
+
+    final uri = _buildUri(
+      url,
+      token: token,
+      options: options,
+    );
 
     try {
-      final ws = await platform.connectToWebSocket(Uri.parse(rtcUrl + params));
+      final ws = await platform.connectToWebSocket(uri);
       ws.stream.listen(_handleMessage, onError: _handleError, onDone: _handleDone);
       _ws = ws;
-    } catch (e) {
-      final completer = Completer<void>();
-      final validateUri = Uri.parse('http${rtcUrl.substring(2)}/validate$params');
-      http.get(validateUri).then((response) {
-        if (response.statusCode != 200) {
-          completer.completeError(ConnectError(response.body));
+    } catch (socketError) {
+      //
+      // Validate mode
+      //
+      final validateUri = _buildUri(
+        url,
+        token: token,
+        options: options,
+        validate: true,
+      );
+
+      //
+      // Attempt Validation
+      //
+      try {
+        final validateResponse = await http.get(validateUri);
+        if (validateResponse.statusCode != 200) {
+          throw ConnectError(validateResponse.body);
         } else {
-          completer.completeError(ConnectError());
+          throw ConnectError();
```

Comment:
so this section is pretty tricky due to the various states it goes through.. I'm not sure why I had used a completer originally (I had changed the implementation here a few times). Hopefully it's no longer needed.

Reply:
After re-checking you were right something was weird so I fixed it, please check next commit. 👍🏻

Reply:
Should be fixed in 73578a497fde6fe1f8ec524eaadd533a23f7622b